cf_units/_udunits2_parser/udunits2Lexer.g4.jinja

```diff
@@ -1,3 +1,156 @@
 lexer grammar udunits2Lexer;
 
-ALL: .;
+// Whitespace is significant, and is not ignored.
+// For example: "m 2" == "m*2", yet "m2" == "m^2".
+
+
+SIGNED_INT : (MINUS|PLUS) INT ;
+
+PLUS: '+' ;
+fragment MINUS: '-' ;
+MULTIPLY: ('*' | '·' | MINUS) ;
+DIVIDE: WS* ('/' | ' PER ' | ' per ') WS*;
+PERIOD: '.' ;
+
+OPEN_PAREN: '(' ;
+CLOSE_PAREN: ')' ;
+SEMICOLON: ':' ;
+
+INT : '0'..'9'+ ;
+
+
+fragment ANY_INT: INT | SIGNED_INT ;
+E_POWER: ('E' | 'e') ANY_INT ;
+FLOAT:
+    // NOTE: floats themselves are context sensitive. (e.g. m2.3 === m^2 * 3 in udunits2)
+    // For this reason, FLOATS are parsed *after* MULTIPLY (which contains '.').
+    // This behaviour is reversed immediate after seeing an ID token.
+    // Example: -2.e5
+    (((ANY_INT PERIOD INT?)
+     |(ANY_INT? PERIOD INT)
+    ) E_POWER?)  // 1.2e-5, 1e2, 2.e4
+    | (ANY_INT E_POWER)
+;
+
+
+SHIFT_OP:
+  ( '@'
+  | 'after'
+  | 'from'
+  | 'since'
+  | 'ref'
+  )  -> pushMode(SHIFT_MODE)
+;
+
+
+UNICODE_EXPONENT:
+  // One or more ISO-8859-9 encoded exponent characters
+  ('⁻' | '⁺' | '¹' | '²' | '³' | '⁴' | '⁵' | '⁶' | '⁷' | '⁸' | '⁹' | '⁰')+
+; 
+
+RAISE :
+  ( '^'
+  | '**'
+  )
+;
+
+LOG: 'log' | 'lg' | 'ln' | 'lb';
+
+LOGREF: '(' WS* ('RE' | 're') ':'? WS*;
+
+//ID: one of
+//        <id>
+//        '%'
+//        "'"
+//        "\""
+//        degree sign
+//        greek mu character
+//
+
+ID:  ([A-Za-z_] | LATIN_SUBSET)+ -> mode(ID_SEEN);
+
+LATIN_SUBSET:
+    '\u00C0'..'\u00D6'  // UDUNITS implementation "\xc3([\x80-\x96])"
+  | '\u00D8'..'\u00F6'  // UDUNITS implementation "\xc3([\x98-\xB6])"
+  | '\u00F8'..'\u00FF'  // UDUNITS implementation "\xc3([\xB8-\xBF])"
+  | '\u0080' | '\u00AD' // Non-breaking space. " ". Why is this in UDUNITS?!
+  | '\u00B0'            // Degree symbol °
+  | '\u00B5'            // Mu µ
+  | 'π' | 'Ω'           // NOTE: Other symbols are allowed, as long as they are defined in the XML.
+;
+
+WS : [ ] ;
+
+// Any characters which fail to match will raise an error.
+// nb. without this token, we don't get quite as good error messages (e.g. "$")
+// TODO: test for this!
```

Comment:
There is a todo here. The tests do actually give you coverage if you remove this element, but they aren't testing the quality of the SyntaxError that is returned. That would be a nice thing to do systematically.